### Importing Necessary Libraries

In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns

### Reading the Dataset

In [2]:
df = pd.read_csv(r'C:\Users\Administrator\Downloads\Lead Scoring Assignment\Leads.csv')

### Inspection of Dataset

In [3]:
# to know the dataset
df.head(3)

Prospect ID  Lead Number              Lead Origin  \
0  7927b2df-8bba-4d29-b9a2-b6e0beafe620       660737                      API   
1  2a272436-5132-4136-86fa-dcc88c88f482       660728                      API   
2  8cc8c611-a219-4f35-ad23-fdfd2656bd8a       660727  Landing Page Submission   

      Lead Source Do Not Email Do Not Call  Converted  TotalVisits  \
0      Olark Chat           No          No          0          0.0   
1  Organic Search           No          No          0          5.0   
2  Direct Traffic           No          No          1          2.0   

   Total Time Spent on Website  Page Views Per Visit  ...  \
0                            0                   0.0  ...   
1                          674                   2.5  ...   
2                         1532                   2.0  ...   

  Get updates on DM Content    Lead Profile    City  \
0                        No          Select  Select   
1                        No          Select  Select   
2                        No  Potential Lead  Mumbai   

  Asymmetrique Activity Index Asymmetrique Profile Index  \
0                   02.Medium                  02.Medium   
1                   02.Medium                  02.Medium   
2                   02.Medium                    01.High   

  Asymmetrique Activity Score Asymmetrique Profile Score  \
0                        15.0                       15.0   
1                        15.0                       15.0   
2                        14.0                       20.0   

  I agree to pay the amount through cheque  \
0                                       No   
1                                       No   
2                                       No   

  A free copy of Mastering The Interview Last Notable Activity  
0                                     No              Modified  
1                                     No          Email Opened  
2                                    Yes          Email Opened  

[3 rows x 37 columns]

In [4]:
# to know the shape of dataset / # of rows and # of columns
df.shape

(9240, 37)

In [5]:
# to know the datatypes and null in every columns
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 9240 entries, 0 to 9239
Data columns (total 37 columns):
 #   Column                                         Non-Null Count  Dtype  
---  ------                                         --------------  -----  
 0   Prospect ID                                    9240 non-null   object 
 1   Lead Number                                    9240 non-null   int64  
 2   Lead Origin                                    9240 non-null   object 
 3   Lead Source                                    9204 non-null   object 
 4   Do Not Email                                   9240 non-null   object 
 5   Do Not Call                                    9240 non-null   object 
 6   Converted                                      9240 non-null   int64  
 7   TotalVisits                                    9103 non-null   float64
 8   Total Time Spent on Website                    9240 non-null   int64  
 9   Page Views Per Visit                           9103 

In [6]:
# for descriptive analysis of the dataset
df.describe()

Lead Number    Converted  TotalVisits  Total Time Spent on Website  \
count    9240.000000  9240.000000  9103.000000                  9240.000000   
mean   617188.435606     0.385390     3.445238                   487.698268   
std     23405.995698     0.486714     4.854853                   548.021466   
min    579533.000000     0.000000     0.000000                     0.000000   
25%    596484.500000     0.000000     1.000000                    12.000000   
50%    615479.000000     0.000000     3.000000                   248.000000   
75%    637387.250000     1.000000     5.000000                   936.000000   
max    660737.000000     1.000000   251.000000                  2272.000000   

       Page Views Per Visit  Asymmetrique Activity Score  \
count           9103.000000                  5022.000000   
mean               2.362820                    14.306252   
std                2.161418                     1.386694   
min                0.000000                     7.000000   
25%                1.000000                    14.000000   
50%                2.000000                    14.000000   
75%                3.000000                    15.000000   
max               55.000000                    18.000000   

       Asymmetrique Profile Score  
count                 5022.000000  
mean                    16.344883  
std                      1.811395  
min                     11.000000  
25%                     15.000000  
50%                     16.000000  
75%                     18.000000  
max                     20.000000

In [7]:
# to know how many null values are present in every columns
df.isnull().sum()

Prospect ID                                         0
Lead Number                                         0
Lead Origin                                         0
Lead Source                                        36
Do Not Email                                        0
Do Not Call                                         0
Converted                                           0
TotalVisits                                       137
Total Time Spent on Website                         0
Page Views Per Visit                              137
Last Activity                                     103
Country                                          2461
Specialization                                   1438
How did you hear about X Education               2207
What is your current occupation                  2690
What matters most to you in choosing a course    2709
Search                                              0
Magazine                                            0
Newspaper Article           

In [8]:
# to know how many null values are present in every columns in terms of percentage
round(df.isnull().sum()/len(df)*100,2)

Prospect ID                                       0.00
Lead Number                                       0.00
Lead Origin                                       0.00
Lead Source                                       0.39
Do Not Email                                      0.00
Do Not Call                                       0.00
Converted                                         0.00
TotalVisits                                       1.48
Total Time Spent on Website                       0.00
Page Views Per Visit                              1.48
Last Activity                                     1.11
Country                                          26.63
Specialization                                   15.56
How did you hear about X Education               23.89
What is your current occupation                  29.11
What matters most to you in choosing a course    29.32
Search                                            0.00
Magazine                                          0.00
Newspaper 

##### **`As we know we can't create a model with null values we need to remove all null values from the dataset`**
##### **`let's start inspecting each column one by one with containing null values`**

In [9]:
df['City'].value_counts()

City
Mumbai                         3222
Select                         2249
Thane & Outskirts               752
Other Cities                    686
Other Cities of Maharashtra     457
Other Metro Cities              380
Tier II Cities                   74
Name: count, dtype: int64

In [10]:
# while inspection I found 'Select' is the common value in most of the column thus it is good to replace all null values for every column with 'Select'

In [11]:
df['Country'].replace(np.nan,'Select', inplace=True)

In [12]:
df['Specialization'].replace(np.nan,'Select', inplace=True)

In [13]:
df['How did you hear about X Education'].replace(np.nan,'Select', inplace=True)

In [14]:
df['What is your current occupation'].replace(np.nan,'Select', inplace=True)

In [15]:
df['What matters most to you in choosing a course'].replace(np.nan,'Select', inplace=True)

In [16]:
df['Tags'].replace(np.nan,'No Comments', inplace=True)

In [17]:
df['Lead Quality'].replace(np.nan,'Not Sure', inplace=True)

In [18]:
df['Lead Profile'].replace(np.nan,'Select', inplace=True)

In [19]:
df['City'].replace(np.nan,'Select', inplace=True)

In [20]:
df.isnull().sum()

Prospect ID                                         0
Lead Number                                         0
Lead Origin                                         0
Lead Source                                        36
Do Not Email                                        0
Do Not Call                                         0
Converted                                           0
TotalVisits                                       137
Total Time Spent on Website                         0
Page Views Per Visit                              137
Last Activity                                     103
Country                                             0
Specialization                                      0
How did you hear about X Education                  0
What is your current occupation                     0
What matters most to you in choosing a course       0
Search                                              0
Magazine                                            0
Newspaper Article           

In [21]:
# here we don't need these columns which contains null value so we can drop them

df.drop(['Asymmetrique Activity Index',
       'Asymmetrique Profile Index', 'Asymmetrique Activity Score',
       'Asymmetrique Profile Score'], axis=1, inplace=True)

**As we know in these column there are very less amount of null values so we can drop that rows which contains null values**
Lead Source                                        36
TotalVisits                                       137
Page Views Per Visit                              137
Last Activity                                     103
**total 413 rows will be deleted from 9240 rows (9240-413 = 8827)**

In [22]:
# this method will drop rows which contains null values
df.dropna(inplace=True)

In [23]:
# to know how many null values are present in every columns in terms of percentage, after handling null values
round(df.isnull().sum()/len(df)*100,2)

Prospect ID                                      0.0
Lead Number                                      0.0
Lead Origin                                      0.0
Lead Source                                      0.0
Do Not Email                                     0.0
Do Not Call                                      0.0
Converted                                        0.0
TotalVisits                                      0.0
Total Time Spent on Website                      0.0
Page Views Per Visit                             0.0
Last Activity                                    0.0
Country                                          0.0
Specialization                                   0.0
How did you hear about X Education               0.0
What is your current occupation                  0.0
What matters most to you in choosing a course    0.0
Search                                           0.0
Magazine                                         0.0
Newspaper Article                             

### Encoding

##### `converting all object columns into numerical columns`

In [24]:
# As we know datatypes for all column should be a number for creating a AI Model
df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 9074 entries, 0 to 9239
Data columns (total 33 columns):
 #   Column                                         Non-Null Count  Dtype  
---  ------                                         --------------  -----  
 0   Prospect ID                                    9074 non-null   object 
 1   Lead Number                                    9074 non-null   int64  
 2   Lead Origin                                    9074 non-null   object 
 3   Lead Source                                    9074 non-null   object 
 4   Do Not Email                                   9074 non-null   object 
 5   Do Not Call                                    9074 non-null   object 
 6   Converted                                      9074 non-null   int64  
 7   TotalVisits                                    9074 non-null   float64
 8   Total Time Spent on Website                    9074 non-null   int64  
 9   Page Views Per Visit                           9074 non-n

In [25]:
# to know all (object and numerical both) distinct value counts of each columns
for cols in df.columns:
    print(cols, ":", df[cols].nunique())

Prospect ID : 9074
Lead Number : 9074
Lead Origin : 4
Lead Source : 21
Do Not Email : 2
Do Not Call : 2
Converted : 2
TotalVisits : 41
Total Time Spent on Website : 1717
Page Views Per Visit : 114
Last Activity : 17
Country : 39
Specialization : 19
How did you hear about X Education : 10
What is your current occupation : 7
What matters most to you in choosing a course : 4
Search : 2
Magazine : 1
Newspaper Article : 2
X Education Forums : 2
Newspaper : 2
Digital Advertisement : 2
Through Recommendations : 2
Receive More Updates About Our Courses : 1
Tags : 27
Lead Quality : 5
Update me on Supply Chain Content : 1
Get updates on DM Content : 1
Lead Profile : 6
City : 7
I agree to pay the amount through cheque : 1
A free copy of Mastering The Interview : 2
Last Notable Activity : 16


In [26]:
# to know distinct value counts of all object columns
# here we can find that which type of encoding we've to be perform in which object column

for cols in df.columns:
    if df[cols].dtype == 'object':
        print(cols, ":", df[cols].nunique())

Prospect ID : 9074
Lead Origin : 4
Lead Source : 21
Do Not Email : 2
Do Not Call : 2
Last Activity : 17
Country : 39
Specialization : 19
How did you hear about X Education : 10
What is your current occupation : 7
What matters most to you in choosing a course : 4
Search : 2
Magazine : 1
Newspaper Article : 2
X Education Forums : 2
Newspaper : 2
Digital Advertisement : 2
Through Recommendations : 2
Receive More Updates About Our Courses : 1
Tags : 27
Lead Quality : 5
Update me on Supply Chain Content : 1
Get updates on DM Content : 1
Lead Profile : 6
City : 7
I agree to pay the amount through cheque : 1
A free copy of Mastering The Interview : 2
Last Notable Activity : 16


#### **`distinct value of 2 are good contender for binary encoding (0 and 1)`**
#### **`distinct value of <10 are good contender for one-hot encoding or dummy variable creation`**
#### **`distinct value of >10 are good contender for frequency encoding`**

In [27]:
# let's start inspecting all elegible columns for binary encoding

In [28]:
df['Newspaper'].value_counts()

Newspaper
No     9073
Yes       1
Name: count, dtype: int64

In [29]:
# df['Do Not Email'].replace({'No':0,'Yes':1}, inplace=True)   #Replacing 'No' with 0 and 'Yes' with 1

In [30]:
# df['Do Not Call'].replace({'No':0,'Yes':1}, inplace=True)   #Replacing 'No' with 0 and 'Yes' with 1

In [31]:
# df['Search'].replace({'No':0,'Yes':1}, inplace=True)   #Replacing 'No' with 0 and 'Yes' with 1

In [32]:
# df['Newspaper Article'].replace({'No':0,'Yes':1}, inplace=True)   #Replacing 'No' with 0 and 'Yes' with 1

In [33]:
# df['X Education Forums'].replace({'No':0,'Yes':1}, inplace=True)   #Replacing 'No' with 0 and 'Yes' with 1

In [34]:
# df['Newspaper'].replace({'No':0,'Yes':1}, inplace=True)   #Replacing 'No' with 0 and 'Yes' with 1

In [35]:
# df['Digital Advertisement'].replace({'No':0,'Yes':1}, inplace=True)   #Replacing 'No' with 0 and 'Yes' with 1

In [36]:
# df['Through Recommendations'].replace({'No':0,'Yes':1}, inplace=True)   #Replacing 'No' with 0 and 'Yes' with 1

In [37]:
# df['A free copy of Mastering The Interview'].replace({'No':0,'Yes':1}, inplace=True)   #Replacing 'No' with 0 and 'Yes' with 1

In [38]:
# Replacing 'No' with 0 and 'Yes' with 1

bin = ['Do Not Email', 'Do Not Call', 'Search', 'Newspaper Article', 'X Education Forums','Newspaper','Digital Advertisement',
       'Through Recommendations', 'A free copy of Mastering The Interview']

# creating a function for binary mapping
def binary_map(x):
    return x.map({'No':0,'Yes':1})

# Applying the function to the bin list
df[bin] = df[bin].apply(binary_map)

In [39]:
for cols in df.columns:
    if df[cols].dtype == 'object':
        print(cols, ":", df[cols].nunique())

# distinct values of 2 has gone now

Prospect ID : 9074
Lead Origin : 4
Lead Source : 21
Last Activity : 17
Country : 39
Specialization : 19
How did you hear about X Education : 10
What is your current occupation : 7
What matters most to you in choosing a course : 4
Magazine : 1
Receive More Updates About Our Courses : 1
Tags : 27
Lead Quality : 5
Update me on Supply Chain Content : 1
Get updates on DM Content : 1
Lead Profile : 6
City : 7
I agree to pay the amount through cheque : 1
Last Notable Activity : 16


In [40]:
# As we know that count of distinct values less than 10 are applicable for dummy variable creation so now we need to encode these columns

dum = pd.get_dummies(df[['Lead Origin','What is your current occupation','What matters most to you in choosing a course',
                         'Lead Quality','Lead Profile','City']], drop_first=True, dtype=int)

# 4-1=3
# 7-1=6
# 4-1=3
# 5-1=4
# 6-1=5
# 7-1=6
# 27 new columns will created

In [41]:
dum

Lead Origin_Landing Page Submission  Lead Origin_Lead Add Form  \
0                                       0                          0   
1                                       0                          0   
2                                       1                          0   
3                                       1                          0   
4                                       1                          0   
...                                   ...                        ...   
9235                                    1                          0   
9236                                    1                          0   
9237                                    1                          0   
9238                                    1                          0   
9239                                    1                          0   

      Lead Origin_Lead Import  What is your current occupation_Housewife  \
0                           0                                          0   
1                           0                                          0   
2                           0                                          0   
3                           0                                          0   
4                           0                                          0   
...                       ...                                        ...   
9235                        0                                          0   
9236                        0                                          0   
9237                        0                                          0   
9238                        0                                          0   
9239                        0                                          0   

      What is your current occupation_Other  \
0                                         0   
1                                         0   
2                                         0   
3                                         0   
4                                         0   
...                                     ...   
9235                                      0   
9236                                      0   
9237                                      0   
9238                                      0   
9239                                      0   

      What is your current occupation_Select  \
0                                          0   
1                                          0   
2                                          0   
3                                          0   
4                                          0   
...                                      ...   
9235                                       0   
9236                                       0   
9237                                       0   
9238                                       1   
9239                                       0   

      What is your current occupation_Student  \
0                                           0   
1                                           0   
2                                           1   
3                                           0   
4                                           0   
...                                       ...   
9235                                        0   
9236                                        0   
9237                                        0   
9238                                        0   
9239                                        0   

      What is your current occupation_Unemployed  \
0                                              1   
1                                              1   
2                                              0   
3                                              1   
4                                              1   
...                                          ...   
9235                                           1   
9236                                           1   
9237                         

In [42]:
# now concat df and newly created columns named dum to a new dataframe and call this newdf
newdf = pd.concat([df, dum], axis=1)

In [43]:
newdf.head(5)

Prospect ID  Lead Number              Lead Origin  \
0  7927b2df-8bba-4d29-b9a2-b6e0beafe620       660737                      API   
1  2a272436-5132-4136-86fa-dcc88c88f482       660728                      API   
2  8cc8c611-a219-4f35-ad23-fdfd2656bd8a       660727  Landing Page Submission   
3  0cc2df48-7cf4-4e39-9de9-19797f9b38cc       660719  Landing Page Submission   
4  3256f628-e534-4826-9d63-4a8b88782852       660681  Landing Page Submission   

      Lead Source  Do Not Email  Do Not Call  Converted  TotalVisits  \
0      Olark Chat             0            0          0          0.0   
1  Organic Search             0            0          0          5.0   
2  Direct Traffic             0            0          1          2.0   
3  Direct Traffic             0            0          0          1.0   
4          Google             0            0          1          2.0   

   Total Time Spent on Website  Page Views Per Visit  ...  \
0                            0                   0.0  ...   
1                          674                   2.5  ...   
2                         1532                   2.0  ...   
3                          305                   1.0  ...   
4                         1428                   1.0  ...   

  Lead Profile_Other Leads Lead Profile_Potential Lead Lead Profile_Select  \
0                        0                           0                   1   
1                        0                           0                   1   
2                        0                           1                   0   
3                        0                           0                   1   
4                        0                           0                   1   

  Lead Profile_Student of SomeSchool City_Other Cities  \
0                                  0                 0   
1                                  0                 0   
2                                  0                 0   
3                                  0                 0   
4                                  0                 0   

  City_Other Cities of Maharashtra  City_Other Metro Cities City_Select  \
0                                0                        0           1   
1                                0                        0           1   
2                                0                        0           0   
3                                0                        0           0   
4                                0                        0           0   

   City_Thane & Outskirts  City_Tier II Cities  
0                       0                    0  
1                       0                    0  
2                       0                    0  
3                       0                    0  
4                       0                    0  

[5 rows x 60 columns]

In [44]:
# now drop these columns because we already created dummy valiables for these columns so we don't need this

newdf.drop(['Lead Origin','What is your current occupation','What matters most to you in choosing a course',
                         'Lead Quality','Lead Profile','City'],axis=1, inplace=True)

In [45]:
for cols in newdf.columns:
    if newdf[cols].dtype == 'object':
        print(cols, ":", newdf[cols].nunique())

# columns for the dummy valiable creation are also missing

Prospect ID : 9074
Lead Source : 21
Last Activity : 17
Country : 39
Specialization : 19
How did you hear about X Education : 10
Magazine : 1
Receive More Updates About Our Courses : 1
Tags : 27
Update me on Supply Chain Content : 1
Get updates on DM Content : 1
I agree to pay the amount through cheque : 1
Last Notable Activity : 16


In [46]:
# here we don't need that columns which has only 1 distinct value so we need to drop those

newdf.drop(['Magazine','X Education Forums','Newspaper','Receive More Updates About Our Courses','Update me on Supply Chain Content',
            'Get updates on DM Content','I agree to pay the amount through cheque'],axis=1, inplace=True)

In [47]:
for cols in newdf.columns:
    if newdf[cols].dtype == 'object':
        print(cols, ":", newdf[cols].nunique())

# columns for only one unique value are also missing now

Prospect ID : 9074
Lead Source : 21
Last Activity : 17
Country : 39
Specialization : 19
How did you hear about X Education : 10
Tags : 27
Last Notable Activity : 16


In [48]:
# here Prospect ID isn't giving any valuable information so we can drop it too
newdf.drop('Prospect ID',axis=1, inplace=True)

In [49]:
for cols in newdf.columns:
    if newdf[cols].dtype == 'object':
        print(cols, ":", newdf[cols].nunique())

# missing Prospect ID

Lead Source : 21
Last Activity : 17
Country : 39
Specialization : 19
How did you hear about X Education : 10
Tags : 27
Last Notable Activity : 16


In [50]:
# Lead Source : 18
# Last Activity : 16
# Country : 34
# Specialization : 19
# How did you hear about X Education : 10
# Tags : 26
# Last Notable Activity : 13

# all these columns are applicable for frequency encoding



In [51]:
newdf['Lead Source'] = newdf['Lead Source'].map(newdf['Lead Source'].value_counts())

In [52]:
newdf['Last Activity'] = newdf['Last Activity'].map(newdf['Last Activity'].value_counts())

In [53]:
newdf['Country'] = newdf['Country'].map(newdf['Country'].value_counts())

In [54]:
newdf['Specialization'] = newdf['Specialization'].map(newdf['Specialization'].value_counts())

In [55]:
newdf['How did you hear about X Education'] = newdf['How did you hear about X Education'].map(newdf['How did you hear about X Education'].value_counts())

In [56]:
newdf['Tags'] = newdf['Tags'].map(newdf['Tags'].value_counts())

In [57]:
newdf['Last Notable Activity'] = newdf['Last Notable Activity'].map(newdf['Last Notable Activity'].value_counts())

In [58]:
for cols in newdf.columns:
    if newdf[cols].dtype == 'object':
        print(cols, ":", newdf[cols].nunique())

# done all frequency encoding

In [59]:
newdf.head(5)

Lead Number  Lead Source  Do Not Email  Do Not Call  Converted  \
0       660737         1753             0            0          0   
1       660728         1154             0            0          0   
2       660727         2543             0            0          1   
3       660719         2543             0            0          0   
4       660681         2868             0            0          1   

   TotalVisits  Total Time Spent on Website  Page Views Per Visit  \
0          0.0                            0                   0.0   
1          5.0                          674                   2.5   
2          2.0                         1532                   2.0   
3          1.0                          305                   1.0   
4          2.0                         1428                   1.0   

   Last Activity  Country  ...  Lead Profile_Other Leads  \
0            640     2296  ...                         0   
1           3432     6491  ...                         0   
2           3432     6491  ...                         0   
3             90     6491  ...                         0   
4            428     6491  ...                         0   

   Lead Profile_Potential Lead  Lead Profile_Select  \
0                            0                    1   
1                            0                    1   
2                            1                    0   
3                            0                    1   
4                            0                    1   

   Lead Profile_Student of SomeSchool  City_Other Cities  \
0                                   0                  0   
1                                   0                  0   
2                                   0                  0   
3                                   0                  0   
4                                   0                  0   

   City_Other Cities of Maharashtra  City_Other Metro Cities  City_Select  \
0                                 0                        0            1   
1                                 0                        0            1   
2                                 0                        0            0   
3                                 0                        0            0   
4                                 0                        0            0   

   City_Thane & Outskirts  City_Tier II Cities  
0                       0                    0  
1                       0                    0  
2                       0                    0  
3                       0                    0  
4                       0                    0  

[5 rows x 46 columns]

In [60]:
# I think 'lead number' and 'lead source' isn't giving any valuable info so we can also drop them

newdf.drop(['Lead Number', 'Lead Source'], axis=1, inplace=True)

In [61]:
newdf.head(5)

Do Not Email  Do Not Call  Converted  TotalVisits  \
0             0            0          0          0.0   
1             0            0          0          5.0   
2             0            0          1          2.0   
3             0            0          0          1.0   
4             0            0          1          2.0   

   Total Time Spent on Website  Page Views Per Visit  Last Activity  Country  \
0                            0                   0.0            640     2296   
1                          674                   2.5           3432     6491   
2                         1532                   2.0           3432     6491   
3                          305                   1.0             90     6491   
4                         1428                   1.0            428     6491   

   Specialization  How did you hear about X Education  ...  \
0            3282                                7086  ...   
1            3282                                7086  ...   
2             399                                7086  ...   
3             202                                 347  ...   
4            3282                                 186  ...   

   Lead Profile_Other Leads  Lead Profile_Potential Lead  Lead Profile_Select  \
0                         0                            0                    1   
1                         0                            0                    1   
2                         0                            1                    0   
3                         0                            0                    1   
4                         0                            0                    1   

   Lead Profile_Student of SomeSchool  City_Other Cities  \
0                                   0                  0   
1                                   0                  0   
2                                   0                  0   
3                                   0                  0   
4                                   0                  0   

   City_Other Cities of Maharashtra  City_Other Metro Cities  City_Select  \
0                                 0                        0            1   
1                                 0                        0            1   
2                                 0                        0            0   
3                                 0                        0            0   
4                                 0                        0            0   

   City_Thane & Outskirts  City_Tier II Cities  
0                       0                    0  
1                       0                    0  
2                       0                    0  
3                       0                    0  
4                       0                    0  

[5 rows x 44 columns]